Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs

```diff
@@ -175,174 +175,336 @@ public string SequenceWithChemicalFormulas
-                NeutralTerminusFragment[] terminalMasses = new CompactPeptide(this, temporaryFragmentationTerminus).TerminalMasses;
-
-                int firstUsableIndex = FullSequence.Length;
-                bool completeFragmentationSeries = true;
-                switch (productType)
+                // n-term fragments
+                if (calculateNTermFragments)
                 {
-                    case ProductType.aStar:
-                    case ProductType.bStar:
-                    case ProductType.yStar:
-                    case ProductType.aDegree:
-                    case ProductType.bDegree:
-                    case ProductType.yDegree:
+                    // get n-term residue mass
+                    if (Residue.TryGetResidue(BaseSequence[r], out Residue residue))
+                    {
+                        nTermMass += residue.MonoisotopicMass;
+                    }
+                    else
+                    {
+                        nTermMass = double.NaN;
+                    }
+
+                    // add side-chain mod
+                    if (AllModsOneIsNterminus.TryGetValue(r + 2, out Modification mod))
+                    {
+                        nTermMass += mod.MonoisotopicMass.Value;
+                    }
+
+                    // handle star and degree ions for low-res CID
+                    if (dissociationType == DissociationType.LowCID)
+                    {
+                        if (BaseSequence[r] == 'R' || BaseSequence[r] == 'K' || BaseSequence[r] == 'N' || BaseSequence[r] == 'Q')
                         {
-                            firstUsableIndex = GetFirstUsableIndex(FullSequence, productType, temporaryFragmentationTerminus);
-                            completeFragmentationSeries = false;
-                        };
-                        break;
+                            haveSeenNTermStarIon = true;
+                        }
 
-                    default:
-                        break;
-                }
+                        if (BaseSequence[r] == 'S' || BaseSequence[r] == 'T' || BaseSequence[r] == 'E' || BaseSequence[r] == 'D')
+                        {
+                            haveSeenNTermDegreeIon = true;
+                        }
+                    }
 
-                for (int f = 0; f < terminalMasses.Length; f++)
-                {
-                    if (completeFragmentationSeries)
+                    // skip first N-terminal fragment (b1, aDegree1, ...) for CID
+                    if (r == 0 && (dissociationType == DissociationType.CID || dissociationType == DissociationType.LowCID))
                     {
-                        // fragments with neutral loss
-                        if (AllModsOneIsNterminus.TryGetValue(terminalMasses[f].AminoAcidPosition + 1, out Modification mod) && mod.NeutralLosses != null
+                        goto CTerminusFragments;
```

Comment:
![image](https://user-images.githubusercontent.com/16883585/69364887-d21d1c80-0c58-11ea-8758-752d8ac851e3.png)

Reply:
wow. i didn't even know that goto still existed. when i learned to program in the 80s, we called program with gotos "spaghetti code"